File: Java/GenerateParentheses.java
import java.util.*;
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generateHelper(n, result, "", 0, 0);
        return result;
    }
    
    private void generateHelper(int n, List<String> result, String current, int left, int right) {
        if (current.length() == 2*n) {
            result.add(current);
            return;
        }
        
        if (left < n) {
            generateHelper(n, result, current + '(', left + 1, right);
        }
        if (right < left) {
            generateHelper(n, result, current + ')', left, right + 1);
        }
        
    }
}

File: Python/GenerateParentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.helper(0, 0, result, "", n)
        return result

    def helper(self, left, right, result, current, n):
        if left + right == 2*n:
            result.append(current)
        if left < n:
            self.helper(left + 1, right, result, current + '(', n)
        if right < left:
            self.helper(left, right + 1, result, current + ')', n)